File: Samyam-Cpp/Lab1/RailFence.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

string railFenceEncrypt(const string& original, int key) {
    string encrypted;
    int i, j = 0, direction = 1;
    vector<vector<char>> rail(key, vector<char>(original.size(), ' '));

    for (i = 0; i < original.size(); i++) {
        rail[j][i] = original[i];
        if (j == key - 1) {
            direction = -1; // Change direction when reaching bottom rail
        } else if (j == 0) {
            direction = 1; // Change direction when reaching top rail
        }
        j += direction;
    }

    // Construct encrypted message from rail
    for (i = 0; i < key; i++) {
        for (j = 0; j < original.size(); j++) {
            if (rail[i][j] != ' ') {
                encrypted += rail[i][j];
            }
        }
    }

    return encrypted;
}

string railFenceDecrypt(const string& encrypted, int key) {
    string decrypted;
    int i, j = 0, direction = 1;
    vector<vector<char>> rail(key, vector<char>(encrypted.size(), ' '));

    // Fill the rail fence structure with placeholders
    for (i = 0; i < encrypted.size(); i++) {
        rail[j][i] = '*'; // Placeholder to mark rail positions
        if (j == key - 1) {
            direction = -1; // Change direction when reaching bottom rail
        } else if (j == 0) {
            direction = 1; // Change direction when reaching top rail
        }
        j += direction;
    }

    // Fill the rail fence with the encrypted message
    int index = 0;
    for (i = 0; i < key; i++) {
        for (j = 0; j < encrypted.size(); j++) {
            if (rail[i][j] == '*') {
                rail[i][j] = encrypted[index++];
            }
        }
    }

    // Read the message back from the rail fence
    j = 0;
    direction = 1;
    for (i = 0; i < encrypted.size(); i++) {
        decrypted += rail[j][i];
        if (j == key - 1) {
            direction = -1; // Change direction when reaching bottom rail
        } else if (j == 0) {
            direction = 1; // Change direction when reaching top rail
        }
        j += direction;
    }

    return decrypted;
}
int main() {
    string message;
    int key;

    // Input message and key
    cout << "Enter the message: ";
    getline(cin, message); // Allowing spaces in the message
    cout << "Enter the key: ";
    cin >> key;

    // Encryption
    string encrypted = railFenceEncrypt(message, key);
    cout << "Encrypted message: " << encrypted << endl;

    // Decryption
    string decrypted = railFenceDecrypt(encrypted, key);
    cout << "Decrypted message: " << decrypted << endl;

    return 0;
}

File: Samyam-Go/Lab1/RailFence.go
package main

// import (
// 	"fmt"
// )

func railFenceEncrypt(original string, key int) string {
	encrypted := ""
	rail := make([][]byte, key)
	for i := range rail {
		rail[i] = make([]byte, len(original))
	}

	row, direction := 0, 1

	for i, char := range original {
		rail[row][i] = byte(char)
		row += direction
		if row == key-1 || row == 0 {
			direction = -direction // Change direction at the edges
		}
	}

	for _, row := range rail {
		for _, ch := range row {
			if ch != 0 {
				encrypted += string(ch)
			}
		}
	}

	return encrypted
}

func railFenceDecrypt(encrypted string, key int) string {
	decrypted := ""
	rail := make([][]byte, key)
	for i := range rail {
		rail[i] = make([]byte, len(encrypted))
	}

	i, j, direction := 0, 0, 1

	for _, char := range encrypted {
		char='*'
		rail[j][i] = byte(char)
		if j == key-1 {
			direction = -1 // Change direction when reaching bottom rail
		} else if j == 0 {
			direction = 1 // Change direction when reaching top rail
		}
		j += direction
		i++
	}

	index := 0
	for i := range rail {
		for j := range rail[i] {
			if rail[i][j] == '*' {
				rail[i][j] = encrypted[index]
				index++
			}
		}
	}

	i, j, direction = 0, 0, 1
	for range encrypted { // Removed char variable, since it's not used
		decrypted += string(rail[j][i])
		if j == key-1 {
			direction = -1 // Change direction when reaching bottom rail
		} else if j == 0 {
			direction = 1 // Change direction when reaching top rail
		}
		j += direction
		i++
	}

	return decrypted
}

// func main() {
// 	var message string
// 	var key int

// 	// Input message and key
// 	fmt.Println("Enter the message: ")
// 	fmt.Scanln(&message)
// 	fmt.Println("Enter the key: ")
// 	fmt.Scanln(&key)

// 	// Encryption
// 	encrypted := railFenceEncrypt(message, key)
// 	fmt.Println("Encrypted message: ", encrypted)

// 	// Decryption
// 	decrypted := railFenceDecrypt(encrypted, key)
// 	fmt.Println("Decrypted message: ", decrypted)
// }
